design/design-doc-composition-functions.md

```diff
@@ -175,32 +200,606 @@ metadata:
   name: example
 spec:
   compositeTypeRef:
-    apiVersion: database.example.org/v1alpha1
+    apiVersion: database.example.org/v1
     kind: XPostgreSQLInstance
-  functions:
-  - name: my-cool-function
-    type: Container
-    container:
-      image: xkpg.io/my-cool-function:0.1.0
+  # This Composition uses a pipeline of Functions instead of (P&T) resources.
+  pipeline:
+    # Each step in the pipeline calls one Composition Function.
+  - step: compose-xr-using-go-templates
+    # The functionRef tells the Composition which Function to call. Crossplane
+    # passes the desired and observed state of the XR and any existing composed
+    # resources as 'arguments' to the Function call.
+    functionRef:
+      name: go-templates
+    # A Function call may optionally accept configuration. Think of this like an
+    # additional, optional argument to the Function call. The configuration is a
+    # nested KRM resource - i.e. it has an apiVersion and kind.
+    config:
+      apiVersion: example.org/v1
+      kind: GoTemplate
+      source: Remote
+      remote: git://github.com/example/my-xpostgresql-go-templates
+    # A pipeline can have multiple steps. Each step is processed in order. This
+    # validation step is passed the desired state accumulated by the prior step.
+  - step: validate-composed-resources
+    functionRef:
+      name: cel-validation
 ```
 
-Under this proposal each function is the entrypoint of an OCI image, though the
-API is designed to support different function implementations (such as webhooks)
-in the future. The updated API would affect only the `Composition` type - no
-changes would be required to the schema of `CompositeResourceDefinitions`, XRs,
-etc.
+The updated API would affect only the `Composition` type - no changes would be
+required to the schema of `CompositeResourceDefinitions`, XRs, etc.
 
-Notably the functions would not be responsible for interacting with the API
-server to create, update, or delete composed resources. Instead, they instruct
-Crossplane which resources should be created, updated, or deleted.
+Notably the Functions would not need to be responsible for interacting with the
+API server to create, update, or delete composed resources. Instead, they
+instruct Crossplane which resources should be created, updated, or deleted.
 
-Under the proposed design functions could also be used for purposes besides
+Under the proposed design Functions could also be used for purposes besides
 rendering composed resources, for example validating the results of the
-`resources` array or earlier functions in the `functions` array. Furthermore, a
-function could also be used to implement 'side effects' such as triggering a
+`resources` array or earlier Functions in the `pipeline`. Furthermore, a
+Function could also be used to implement 'side effects' such as triggering a
```

Comment:
While we cannot stop this from happening, it feels odd to call this out as an expected (and encouraged?) practice.

Reply:
This is copied from the original design. It's a side-effect of @bassam pushing us to consider whether Functions should be used for "day two" operations. I don't think anyone feels great about that - I can remove it.